### PR TITLE
Aggregate Table for Miners, Sharders, Authorizers and Validators

### LINKED ISSUE #1906: Create aggregate table for each provider
We need similar to AggregateBlobber table but for authorizer, miner, sharder, validator.

### PR DESCRIPTION
**Pull request by K-Kumar-01**
Signed-off-by: Kushal Kumar <kushalkumargupta4@gmail.com>

## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

File: code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+// nolint
+func (edb *EventDb) updateAuthorizerAggregate(round, pageAmount int64, gs *globalSnapshot) {
+	count, err := edb.GetAuthorizerCount()
+	if err != nil {
+		logging.Logger.Error("update_authorizer_aggregates", zap.Error(err))
+		return
+	}
+	size, currentPageNumber, subpageCount := paginate(round, pageAmount, count, edb.PageLimit())
+
+	exec := edb.Store.Get().Exec("CREATE TEMP TABLE IF NOT EXISTS temp_ids "+
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate.go]: this table will collude with one in blobber aggregates

File: code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+	oldAuthorizers, err := edb.getAuthorizerSnapshots(limit, offset)
+	if err != nil {
+		logging.Logger.Error("getting Authorizer snapshots", zap.Error(err))
+		return
+	}
+	logging.Logger.Debug("Authorizer_snapshot", zap.Int("total_old_Authorizers", len(oldAuthorizers)))
+
+	var aggregates []AuthorizerAggregate
+	for _, current := range currentAuthorizers {
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate.go:132]: use common code to process Provider fields, avoid copy paste

File: code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+		aggregate.UnstakeTotal = (old.UnstakeTotal + current.UnstakeTotal) / 2
+		aggregate.Fee = (old.Fee + current.Fee) / 2
+		aggregate.ServiceCharge = (old.ServiceCharge + current.ServiceCharge) / 2
+		aggregates = append(aggregates, aggregate)
+
+		gs.totalTxnFees += aggregate.Fee
+
+		//gs.totalWritePricePeriod += aggregate.WritePrice
+
+		//gs.AverageTxnFee = .
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate.go]: remove commented code

File: code/go/0chain.net/smartcontract/dbs/event/miner.go [hunk 4]
```diff
@@ -30,6 +30,9 @@ type Miner struct {
 	Active          bool
 	Longitude       float64
 	Latitude        float64
+
+	// number of transactions involved
+	Count int
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/dbs/event/miner.go]: don't add count here

**Review by dabasov — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+		return
+	}
+
+	for i := 0; i < subpageCount; i++ {
+		edb.calculateAuthorizerAggregate(gs, round, edb.PageLimit(), int64(i)*edb.PageLimit())
+	}
+}
+
+// nolint
+func (edb *EventDb) calculateAuthorizerAggregate(gs *globalSnapshot, round, limit, offset int64) {
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate.go:97]: add this method call to worker

Comment (K-Kumar-01) [reply code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate.go:97]: What exactly do you mean by above? I was seeing where `calculateBlobberAggregate` was added to worker for reference, but was unable to find one.

File: code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate.go (hunk 2 above)
Comment (K-Kumar-01) [reply code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate.go:132]: The for loop is similar but the fields vary, so how to use the `common code`.

File: code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate.go (hunk 1 above)
Comment (K-Kumar-01) [reply code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate.go:52]: @dabasov 
Not understanding what is meant by this and how to resolve it.

Comment (K-Kumar-01) [reply code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate.go]: Should I rename the table names them?

Comment (dabasov) [reply code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate.go]: yep

File: code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate.go (hunk 5 above)
Comment (dabasov) [reply code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate.go:97]: strange, check here
https://github.com/0chain/0chain/blob/b5ff530fb9a3a7135b77d4b9b8f30ea761b2d3b3/code/go/0chain.net/smartcontract/dbs/event/blobber_aggregate.go#L72

File: code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate.go (hunk 2 above)
Comment (dabasov) [reply code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate.go:132]: actually fields are the same, but on different structures, we can abstract them out by the interface for example, 

```go
type ProviderAggregate Interface {
    SetTotalStake(newVal int64)
    SetUnstakeTotal(newVal int64)
    ...
    TotalStake() int64
    TotalUnstake() int64
}
```
then we can common code that takes 2 snapshots and assigns one

```go
recalculateProviderFields(new, old ProviderAggregates) {
    new.SetTotalStake(old.TotalStake())
  ....
}

```

File: code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate.go (hunk 5 above)
Comment (K-Kumar-01) [reply code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate.go:97]: @dabasov 
Actually, I am using `edb. calculateAuthorizerAggregate` as [here].(https://github.com/0chain/0chain/blob/409c5bd43982f4f0e930bb6b3c4bcaf3d1e4fca3/code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate.go#L61)

File: code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate.go (hunk 2 above)
Comment (dabasov) [reply code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate.go:132]: UPD: we can do it better with generics, if you need to return new aggregate from recalculate

```go
recalculateProviderFields[T ProviderAggregates](new, old T) T {
    new.SetTotalStake(old.TotalStake())
  ....
    return newAggregate
}
```

**Review by dabasov — APPROVED** [review]
lgtm

**Review by dabasov — APPROVED** [review]
lgtm

**Pull request merged**